Issue 1475/1590 Migration: Tree/Model/Viewer

dleszcz opened the pull request:
#### Description

- Tree service migration
- Remove old code

simplypixi commented:
@carmenfan Whoooah, it's very laggy... Is this model bigger than `Lego federation`? Do we have access to this model on the development environment (database)?

> Can we change the parent of invisible back to a lighter blue colour? grey kind of makes it look as if it's invisible.

Yes, we can. 

> The menu should only be available on hover:

Did you mean `on click`? If not, what about touch devices?

> Instead of having this weird conditions of only being able to see the link button when it's expanded, can we change it so that the link button replaces the return to top? (they should all only be shown when on hover)

Ok! ;)

carmenfan commented:
> @carmenfan Whoooah, it's very laggy... Is this model bigger than Lego federation? Do we have access to this model on the development environment (database)?

I don't think it's that much bigger than lego fed...

I've uploaded it onto your database, in `carmen` -> `Clinic`. The one i used was `Clinic fed (no mep)`, `Clinic fed (full)` has pipes in it so it'll be even bigger.

I had to change the password of `carmen` so i can login to do this 😆 


> > The menu should only be available on hover:
> 
> Did you mean `on click`? If not, what about touch devices?
> 

Can we do both? hover it appears, if click it stays until something else was clicked.

dleszcz commented:
- [ ]  > * If i hide then unhide a model with IFC spaces, it should return to it's original state (`parent of invisible`) not `visible`

- [ ]  > * Highlights and selections are not being shown on the tree:

- [ ]  > * The menu should only be available on hover:

- [ ]  > * Can we change the parent of invisible back to a lighter blue colour? grey kind of makes it look as if it's invisible.

- [ ]  > * The link needs to set with `noopener` or it'll crash both tabs as soon as unity loads due to out of memory

- [ ]  > * Instead of having this weird conditions of only being able to see the link button when it's expanded, can we change it so that the link button replaces the return to top?  (they should all only be shown when on hover)

simplypixi commented:
@carmenfan I pushed some optimizations (I got rid of the processing worker - it was spending a lot of time on serializing the data) and I fixed IFC issue. Could you check how it works on your side?

carmenfan commented:
> 
> 
> @carmenfan I pushed some optimizations (I got rid of the processing worker - it was spending a lot of time on serializing the data) and I fixed IFC issue. Could you check how it works on your side?

@simplypixi 
Much better, but still problematic in some cases:
- I can't tell if it's just out right not working or it's super slow (as I think I've seen it work sometimes), selecting the background seems to take a long time/doesn't unselect the objects.
- Selecting multiple objects (easiest example is to click on the root), still takes forever:
![2](https://user-images.githubusercontent.com/11945337/58881911-1597c480-86d3-11e9-88e8-d6c7e0c1cbd9.gif)
- Whilst I was making screenshots for some of the bugs below, I felt that selection and isolation were very laggy




Other bugs found:
- [ ] Clicking on an object on the viewer, the tree doesnt' scroll to the right place (left is staging, right is branch)
![image](https://user-images.githubusercontent.com/11945337/58882106-758e6b00-86d3-11e9-8020-0245a16362f6.png)
- [ ] Double scroll bar appearing
![image](https://user-images.githubusercontent.com/11945337/58882209-a373af80-86d3-11e9-93c2-2c56b48a89fc.png)
- 2 things here:
    - [ ] Selection box only seems to appear if I click on a leaf node
    - [ ] Parent of selected is not being coloured correctly if you select the parent itself
![2](https://user-images.githubusercontent.com/11945337/58882370-036a5600-86d4-11e9-9cd8-8f8cebc024fc.gif)

- [ ] If I selected something, and hidden it, the item is no longer highlighted on the tree, but the parents are still being highlighted even though there's nothing selected:
![2](https://user-images.githubusercontent.com/11945337/58882533-56dca400-86d4-11e9-8542-1a552f31ca52.gif)

- [ ] As mentioned somewhere in this thread, please make the hover menu stay if the object is currently selected. (selected as in there is a rectangle around the entry.) This includes hidden objects.

- [ ] Clicking Isolate should not unhide an IFC space object unless `Hide IFC Spaces` has been unchecked:
![2](https://user-images.githubusercontent.com/11945337/58882798-03b72100-86d5-11e9-8881-d8decf5d1f54.gif)

- [ ] Ctrl and Alt alters the cursor on the viewer, that should be the same on the tree as you can multi select on the tree.
![2](https://user-images.githubusercontent.com/11945337/58882981-72947a00-86d5-11e9-80c7-c6c579bcfeae.gif)

- [ ] Only the last selected should have a rectangle over it
![2](https://user-images.githubusercontent.com/11945337/58883122-af607100-86d5-11e9-91d6-10fa15b4c7d8.gif)

- [ ] If I selected multiple objects then hit hide/isolate, it should hid/isolate them all (Note: this was really slow to operate):
![2](https://user-images.githubusercontent.com/11945337/58883482-6d83fa80-86d6-11e9-94da-ccc7fb0ca4b6.gif)

simplypixi commented:
> * If I selected multiple objects then hit hide/isolate, it should hid/isolate them all (Note: this was really slow to operate):
>   ![2](https://user-images.githubusercontent.com/11945337/58883482-6d83fa80-86d6-11e9-94da-ccc7fb0ca4b6.gif)

Why? :O Does it make sense? What about hide/isolate button on the toolbar?
These buttons have a totally different context (local) than buttons from the toolbar (global). They are defined per row, so the action should be related only to the row.
What about `scroll to top` if the user selects nodes from different parent nodes? In this case, the action should behave in the same way as actions of other buttons in a row, so the container should be scroll to the parent of selected nodes, which is impossible because they have different parents. xD

Let me show you another example.
Online store with fancy things. You selected a few items (using e.g checkbox). Option `add all to cart` had shown up, but then you clicked 'add to cart' button, which belongs to a single item. What happens? :)

If you want I can add here hiding/isolating selected items, but I think is illogical from user perspective.

paloknapo commented:
> > * If I selected multiple objects then hit hide/isolate, it should hid/isolate them all (Note: this was really slow to operate):
> >   ![2](https://user-images.githubusercontent.com/11945337/58883482-6d83fa80-86d6-11e9-94da-ccc7fb0ca4b6.gif)
> 
> Why? :O Does it make sense? What about hide/isolate button on the toolbar?
> These buttons have a totally different context (local) than buttons from the toolbar (global). They are defined per row, so the action should be related only to the row.
> What about `scroll to top` if the user selects nodes from different parent nodes? In this case, the action should behave in the same way as actions of other buttons in a row, so the container should be scroll to the parent of selected nodes, which is impossible because they have different parents. xD
> 
> Let me show you another example.
> Online store with fancy things. You selected a few items (using e.g checkbox). Option `add all to cart` had shown up, but then you clicked 'add to cart' button, which belongs to a single item. What happens? :)
> 
> If you want I can add here hiding/isolating selected items, but I think is illogical from user perspective.

Good question, it all depends on context though. In our case, we are developing a tool for professionals in construction industry and mental model we adopted here is from CAD applications (AutoCAD, Rhino, SketchUp...) familiar to our users. It might seem illogical to a "shopping user" but makes complete sense to a "CAD user".

Scroll to top is mainly for federations and covers a use case, where I selected object in long list and don't know to which model (discipline/consultant) it belongs to, hence I want to see the top/model level. In case of two or more selected objects in different models, collapse to model level and highlight relevant models.

simplypixi commented:
> > > * If I selected multiple objects then hit hide/isolate, it should hid/isolate them all (Note: this was really slow to operate):
> > >   ![2](https://user-images.githubusercontent.com/11945337/58883482-6d83fa80-86d6-11e9-94da-ccc7fb0ca4b6.gif)
> > 
> > 
> > Why? :O Does it make sense? What about hide/isolate button on the toolbar?
> > These buttons have a totally different context (local) than buttons from the toolbar (global). They are defined per row, so the action should be related only to the row.
> > What about `scroll to top` if the user selects nodes from different parent nodes? In this case, the action should behave in the same way as actions of other buttons in a row, so the container should be scroll to the parent of selected nodes, which is impossible because they have different parents. xD
> > Let me show you another example.
> > Online store with fancy things. You selected a few items (using e.g checkbox). Option `add all to cart` had shown up, but then you clicked 'add to cart' button, which belongs to a single item. What happens? :)
> > If you want I can add here hiding/isolating selected items, but I think is illogical from user perspective.
> 
> Good question, it all depends on context though. In our case, we are developing a tool for professionals in construction industry and mental model we adopted here is from CAD applications (AutoCAD, Rhino, SketchUp...) familiar to our users. It might seem illogical to a "shopping user" but makes complete sense to a "CAD user".

@paloknapo @carmenfan  I've started implementing this as you suggested. I checked your staging environment to be sure how it should work and... it works in the same way as we implemented. So, is it a change?

> Scroll to top is mainly for federations and covers a use case, where I selected object in long list and don't know to which model (discipline/consultant) it belongs to, hence I want to see the top/model level. In case of two or more selected objects in different models, collapse to model level and highlight relevant models.

What if these parent models are on different levels?

carmenfan commented:
> @paloknapo @carmenfan I've started implementing this as you suggested. I checked your staging environment to be sure how it should work and... it works in the same way as we implemented. So, is it a change?

Yes it is a change

> I understand, the tools, which your users use, work that way. But are you sure they like it?
This solution is not flexible at all. Eg. what if someone will decide to hide single node, while some of them will be selected? How can the user achieve that effect?
And what about the habits of current users?

@paloknapo  is a lot more closer to the users so he's best to answer this

> What if these parent models are on different levels?
Shouldn't the same functionality work in the same way on each case?
The logic should be  (Correct me if i'm wrong @paloknapo ):
if federation -> go to root of submodel, if multiple models, collapse the whole tree (or we can just collapse anyway)
if not federation -> go to root of model.

simplypixi commented:
> The logic should be (Correct me if i'm wrong @paloknapo ):
> if federation -> go to root of submodel, if multiple models, collapse the whole tree (or we can just collapse anyway)
> if not federation -> go to root of model.
> 
> I think the confusion is that you were going to the parent of the currently selected, which is not what Pavol wanted.

Makes sense now ;)

simplypixi commented:
@carmenfan I pushed a package of fixes (fixed issues are checked on the list) ;)

simplypixi commented:
@carmenfan All bugs fixed. Let me know if everything works good - I'll rebase the branch.

simplypixi commented:
@carmenfan @paloknapo  I have a small question. Do you really need these buttons in so many places? (see image below)
What do you think about removing the menu and moving button (for hiding ifc spaces) to the toolbar?
![image](https://user-images.githubusercontent.com/5683490/59486782-cf80f480-8e7a-11e9-87d5-ce9607f928cf.png)

simplypixi commented:
@carmenfan Fixed and rebased ;)

carmenfan commented:
> 
> 
> @carmenfan @paloknapo I have a small question. Do you really need these buttons in so many places? (see image below)
> What do you think about removing the menu and moving button (for hiding ifc spaces) to the toolbar?
> ![image](https://user-images.githubusercontent.com/5683490/59486977-75ccfa00-8e7b-11e9-932b-92660d12b899.png)

If anything I don't think we need it on the tree (but @paloknapo wanted it there, so maybe he can give feed back on what he thinks once he's tried it)

The reason why we want it on the bottom tool bar is that the user can interact with the model without the need to pull out the tree tab. I.e. if I'm dealing with issues, and I want to hide some objects I see on my screen, I can highlight it and click hide on the bottom tool bar, without interfering with what i have opened on the left menu

simplypixi commented:
> * Parents of highlighted is not being displayed correctly if I selected my objects via rectanglular highlighting (This seems to also be the case if I selected from the groups component). But it works fine if I selected it without rectangular select enabled. Are they going through different functions? Surely there should only be one function to call to highlight in the tree service.

This is one function but gets some params e.g skipping children etc. Probably this causes the issue.|

>  Isolate button on the tree only works if it was previously selected. The logic should be whatever that has been previously selected (if multi select) and itself.

I think I don't understand. :( The screenshot is (probably) incorrect and maybe that makes me confused a bit.

carmenfan commented:
> 
> I think I don't understand. :( The screenshot is (probably) incorrect and maybe that makes me confused a bit.
If you click on the isolate button on any tree item it's not isolating it.

simplypixi commented:
@carmenfan Fixed ;) 🤞

carmenfan commented on frontend/modules/tree/treeProcessing/processing.ts:
```diff
@@ -0,0 +1,431 @@
+					this.visibilityMap[node._id] = ifcSpacesHidden ? this.defaultVisibilityMap[node._id] : VISIBILITY_STATES.VISIBLE;
+				}
+
+				if (this.isVisibleNode(node._id)) {
```
Assuming we dumped L113. Then we already know the state of the object because we set it.

So maybe we should set it to a variable and use it instead of calling a function to probe the value again?

simplypixi replied:
I added this function to make the code more "readable" - it contains a regular value check, that's all. It's also more convenient in case if you decide e.g to change the data structure.

Nevertheless, I refactored that part of code a bit.

carmenfan commented on frontend/modules/tree/treeProcessing/processing.ts:
```diff
@@ -0,0 +1,431 @@
+			const shouldBeVisible = nodesIds.includes(node._id);
+			if (shouldBeVisible) {
+				if (ifcSpacesHidden) {
+					this.visibilityMap[node._id] = ifcSpacesHidden ? this.defaultVisibilityMap[node._id] : VISIBILITY_STATES.VISIBLE;
```
By the virtue of L113, the else clause will never happen.

Do we actually want L113? because if I'm trying to isolate a node that is currently hidden it wouldn't be be changed to visible?

simplypixi replied:
It's not necessary 👍

simplypixi replied:
✅

carmenfan commented on frontend/modules/tree/treeProcessing/processing.ts:
```diff
@@ -0,0 +1,431 @@
+		}
+
+		if (parents.length) {
+			const uniqeParents = uniqBy(parents, ({ _id}) => _id);
```
```suggestion
			const uniqueParents = uniqBy(parents, ({ _id}) => _id);
```

simplypixi replied:
✅

simplypixi commented on frontend/modules/compare/compare.sagas.ts:
```diff
@@ -286,8 +282,7 @@ function* startComparisonOfFederation() {
 			if (canReuseModel) {
-				const { teamspace, name } = sharedRevisionModel;
-				changeModelNodesVisibility(`${teamspace}:${name}`, true);
+				yield changeModelNodesVisibility([sharedRevisionModel], true);
```
1. Tbh, we need to check it. Probably we wanted to avoid calling some extra logic.
2. Good point! I'll try to do that this way.

simplypixi replied:
✅

simplypixi commented on frontend/modules/tree/tree.selectors.ts:
```diff
@@ -16,14 +16,279 @@
+	}
+);
+
+export const getSelectNodesByIds = (nodesIds) => createSelector(
```
tbh, we used that naming to emphasize that the function returns selector. :P

carmenfan replied:
I kind of figured it may be that. But it's kind of confusing due to select has a special meaning in tree.

I think it may be better to make it as a prefix before the get? I. E. selectGetNodesById

simplypixi replied:
👍

simplypixi replied:
✅

simplypixi commented on frontend/modules/tree/treeProcessing/processing.ts:
```diff
@@ -0,0 +1,431 @@
+	}
+
+	public get selectedNodesIds() {
+		return keys(pickBy(this.selectionMap, (selectionState) => {
```
> I'm guessing the difference between them is one returns only selected nodes, and one returns parents of selected as well as selected?
Yes, exactly.

> maybe highlightedNodesIds should be called fullySelectedIds?
The idea was to set that name because it represents the current state of viewer - nodes returned by that getter are highlighted on the viewer. :) `selected` means they are selected in some way e.g on the nodes list.
But yes, We can change the name. No problem ;)

simplypixi replied:
✅

simplypixi commented on frontend/modules/tree/treeProcessing/processing.ts:
```diff
@@ -0,0 +1,431 @@
+		const children = nodes.map((node) => this.getDeepChildren(node)) as any;
+		const nodesWithChildren = [...nodes, ...children.flat()];
+
+		for (let index = 0; index < filteredNodesIds.length; index++) {
```
Yep, good point! ;)

simplypixi replied:
✅

simplypixi commented:
> * Tree highlight goes away even though the model is still highlighted when you click isolate:

@carmenfan Should I remove the highlight from the model whether add highlighting on the list? Currently, on staging, highlights disappear on both - I suppose, it's the best option, isn't it? Cause the experience won't change from the user perspective.

carmenfan commented:
> > * Tree highlight goes away even though the model is still highlighted when you click isolate:
> 
> @carmenfan Should I remove the highlight from the model whether add highlighting on the list? Currently, on staging, highlights disappear on both - I suppose, it's the best option, isn't it? Cause the experience won't change from the user perspective.

Yea that would be fine.

simplypixi commented:
@carmenfan 
While testing fix for the Issues I noticed one thing. Currently, after issue save, the name has a extra margin ath the bottom. Is it on purpose? Tbh, it looks like a bug. :)

![image](https://user-images.githubusercontent.com/5683490/60085416-8dd53100-9739-11e9-9933-b3d656326067.png)

And your code change:

![image](https://user-images.githubusercontent.com/5683490/60085337-64b4a080-9739-11e9-9b35-ef8eff479c00.png)

Let me know if I should change that. ;)

carmenfan commented:
> @carmenfan
> While testing fix for the Issues I noticed one thing. Currently, after issue save, the name has a extra margin ath the bottom. Is it on purpose? Tbh, it looks like a bug. :)
> 
> ![image](https://user-images.githubusercontent.com/5683490/60085416-8dd53100-9739-11e9-9933-b3d656326067.png)
> 
> And your code change:
> 
> ![image](https://user-images.githubusercontent.com/5683490/60085337-64b4a080-9739-11e9-9b35-ef8eff479c00.png)
> 
> Let me know if I should change that. ;)

If I remember correctly that was added in so we can have multi-line titles. So you can fix it, but make sure if someone type in a very long issue name it still display correctly :)

simplypixi commented on frontend/modules/compare/compare.sagas.ts:
```diff
@@ -239,28 +240,23 @@ function* setTargetModel({ modelId, isTarget, isTypeChange = false }) {
+		return _id;
+	});
+
+	yield changeModelNodesVisibility(nodesIds, isVisible);
```
✅

simplypixi commented on frontend/modules/groups/groups.sagas.ts:
```diff
@@ -148,34 +139,25 @@ export function* selectGroup({ group = {} }) {
+
+			const nodes = yield select(getSelectNodesByIds(nodesIds));
+			const filteredNodes = nodes.filter((n) => n !== undefined);
+			const modelsMap = yield select(getSelectMeshesByNodes(filteredNodes));
```
> Instead of calling
> - getSelectNodesIdsFromSharedIDs
> - getSelectNodesByIds
> - getSelectMeshesByNodes
> 
> Can we make a function to just getMeshesBySharedIds?

Do you mean merging those into one selector?

carmenfan replied:
kind of. Yes make a new selector, but I don't think you need to do `getSelectNodesByIds` because you shouldn't need the node object to get mesh via `idToMeshes` mapping.

simplypixi replied:
How the new selector should work? How can I distinguish what it should return (meshes vs nodes)? I think it over complicates the logic. ;)

Yes, I need these nodes to get the `modelId` and `teamspace`. You wanted us to create something like namespace (`${teamspace}@${modeId}`) to handle connection between nodes and meshes (I think it was implemented this way before), so we did it, but now we have to get somehow this information. ;)
I suppose your idea makes sense only if we get rid of the namespace thing, and use model's id instead of it.
Please look at `getSelectMeshesByIds` in `tree.selectors.ts`.

simplypixi commented on frontend/modules/risks/risks.sagas.ts:
```diff
@@ -328,87 +328,88 @@ const getRiskGroup = async (risk, groupId, revision) => {
+			}
+		} else {
+			const hasViewpointDefaultGroup = risk.viewpoint.group_id;
+			const groupId = hasViewpointDefaultGroup ? risk.viewpoint.group_id : risk.group_id;
```
✅

simplypixi commented on frontend/modules/tree/tree.redux.ts:
```diff
@@ -16,35 +16,140 @@
+	return { ...state, ...auxiliaryMaps };
+};
+
+const setExpanedNodesMap = (state = INITIAL_STATE, { expandedNodesMap }) => {
```
✅

simplypixi commented on frontend/modules/tree/tree.sagas.ts:
```diff
@@ -14,42 +14,198 @@
-			TreeService.nodesClickedBySharedIds(object.selectedNodes);
-			dispatch(TreeActions.getSelectedNodes());
+		Viewer.unhighlightObjects({
+			account: teamspace,
```
Yes, we know about that, but we will change/update Viewer methods on branch with Viewer migration. ;)

simplypixi commented on frontend/modules/tree/tree.sagas.ts:
```diff
@@ -14,42 +14,198 @@
+			const filterdMeshes = meshes.filter((mesh) => nodesSelectionMap[mesh] === SELECTION_STATES.SELECTED);
+			if (meshes.length > 0) {
+				Viewer.highlightObjects({
+					account: teamspace,
```
like above

simplypixi commented on frontend/modules/tree/tree.sagas.ts:
```diff
@@ -14,42 +14,198 @@
+		]);
+		yield put(TreeActions.updateDataRevision());
+	}
+	return false;
```
no, fixed ✅

simplypixi commented on frontend/modules/tree/tree.sagas.ts:
```diff
@@ -14,42 +14,198 @@
+	}
+
+	if (removeGroup) {
+		const nodes = yield select(getSelectNodesByIds(nodesIds));
```
According to another comment, I changed the name to `getSelectNodesbyIds`, because the function returns selector.

simplypixi commented on frontend/modules/tree/tree.sagas.ts:
```diff
@@ -14,42 +14,198 @@
+	if (removeGroup) {
+		const nodes = yield select(getSelectNodesByIds(nodesIds));
+		const activeMeta = yield select(selectActiveMeta);
+		const shouldCloseMeta = nodes.some(({ meta }) => meta.includes(activeMeta));
```
Makes sense ;)

simplypixi replied:
✅

simplypixi commented on frontend/modules/tree/tree.sagas.ts:
```diff
@@ -62,48 +218,336 @@ export function* getSelectedNodes() {
-		Viewer.off(VIEWER_EVENTS.OBJECT_SELECTED);
-		Viewer.off(VIEWER_EVENTS.MULTI_OBJECTS_SELECTED);
-		Viewer.off(VIEWER_EVENTS.BACKGROUND_SELECTED);
+		const nodesIds = yield select(selectInvisibleNodesIds);
```
It can't be named this way. All selector's name should start with `select` - it's a common approach. If you still want to do that this way, we'll also need to change the names of the rest of selectors in the project. ;)

carmenfan replied:
the comment here is more about it should be `hidden` instead of `invisible` :P

simplypixi replied:
Ohh :D So it's fixed :P

simplypixi commented on frontend/modules/tree/tree.sagas.ts:
```diff
@@ -62,48 +218,336 @@ export function* getSelectedNodes() {
+	yield isolateNodes(nodesIds);
+}
+
+function* isolateNode(id) {
```
✅

simplypixi commented on frontend/modules/tree/tree.sagas.ts:
```diff
@@ -62,48 +218,336 @@ export function* getSelectedNodes() {
+/**
+ * SELECT NODES
+ */
+function* selectNode({ id }) {
```
✅

simplypixi commented on frontend/modules/tree/tree.sagas.ts:
```diff
@@ -62,48 +218,336 @@ export function* getSelectedNodes() {
+			yield expandToNode(lastNodeId);
+		}
+		if (nodesIds.length === 1) {
+			yield put(TreeActions.setActiveNode(nodesIds[0]));
```
✅

simplypixi commented on frontend/modules/tree/tree.sagas.ts:
```diff
@@ -62,48 +218,336 @@ export function* getSelectedNodes() {
+	yield put(TreeActions.setExpanedNodesMap(expandedNodesMap));
+}
+
+function* goToParentNode({ nodeId }) {
```
✅

simplypixi commented on frontend/modules/tree/treeProcessing/processing.ts:
```diff
@@ -0,0 +1,431 @@
+		}));
+	}
+
+	public get invisibleNodesIds() {
```
✅

simplypixi commented on frontend/modules/tree/treeProcessing/processing.ts:
```diff
@@ -0,0 +1,431 @@
+	public clearCurrentlySelected = () => {
+		const selectedNodesIds = this.selectedNodesIds;
+		for (let index = 0, size = selectedNodesIds.length; index < size; index++) {
+			this.selectionMap[selectedNodesIds[index]] = SELECTION_STATES.UNSELECTED;
```
✅

simplypixi replied:
Selectors are caching the data using shallow compare of theirs args e.g

```
export const selectFullySelectedNodesIds = createSelector(
	selectTreeProccessing, selectDataRevision, (treeProcessingData) => treeProcessingData.fullySelectedNodesIds || []
);
```
This selector will call `fullySelectedNodesIds` getter only if the reference of `selectTreeProccessing` or the value of `selectDataRevision` will have changed. Reference of `selectTreeProccessing` never changes, so the selector will return "fresh" data only if it will get new data revision.

I think the only thing that we can do yet is splitting `selectionMap` and `visibilityMap` into smaller, separate maps e.g `selectedNodes`, `parentOfUnselectedNodes`, 'visibleNodes`, `parentOfInvisibleNodes`, The downside of this solution is that we will need to rewrite each place which contains these maps.

Let me know if I've explained something unclearly and what you think about that. ;)

simplypixi commented:
@carmenfan 

1. I fixed the name issue.

2. I went through all the bugs and your review. I pushed a lot of fixes and left some comments. Please set conversions as resolved if everything looks good now.

3.
> * Zoom to object on tree only works if you select more than one object, otherwise it seems to zoom out to the whole model (which suggest you may not be passing any ids when you call the function?)

This is weird. Zooming function doesn't get any param, so I also checked how the highlighting function looks like and... it looks good. :( I compared it with results from staging (see below)

###### Tree branch
![image](https://user-images.githubusercontent.com/5683490/60097170-21672b80-9753-11e9-95df-d7f1b555ee3f.png)

###### Staging branch
![image](https://user-images.githubusercontent.com/5683490/60097254-4bb8e900-9753-11e9-8a39-a51d098a2e92.png)

Do you have any idea what is wrong?

carmenfan commented on frontend/modules/compare/compare.sagas.ts:
```diff
@@ -303,6 +294,7 @@ function* startComparisonOfFederation() {
 		}
 	}
 
+	yield setModelsNodesVisibility(modelsToReuse, VISIBILITY_STATES.VISIBLE);
```
not quite what I meant... I think we need 2 arrays here. `modelsToShow`, `modelsToHide`

modelsToShow initialises with baseModels, ditch the calls on L267 L268
while we loop, we add the individual target models to `modelsToShow` or `modelsToHide` depending on whether we can reuse them.

at this point, we call setModelsNodesVisibility on the 2 arrays. Then we reduce this down to 2 calls to set visibility without switching off and on the ones we are to reuse.

I think this makes sense (in my head at least), but maybe I'm missing something?

simplypixi replied:
Done ;)

carmenfan commented:
@Charence  Can you also do some testing on this please?

simplypixi commented:
@carmenfan I pushed the fixes. ;)

dleszcz commented:
> Selection isn't indicated when a filtered tree is shown (no blue highlighting)

Fixed

> * Can we also have a button in the drop down menu to "Select All" items in the tree (only filtered ones if there's a filter applied)?

Added

> I'm getting out-of-memory problems when loading larger models

We check it

simplypixi commented:
> 1. I'm getting out-of-memory problems when loading larger models

@Charence Do we have access to that model? If no, could you upload the model to account named `charence`? I have one idea about what is wrong but I need a proper model to verify it. I think `Skanska federation` is the biggest model, which we have access to, but everything looks good for it.

carmenfan commented:
> 
> 
> > 1. I'm getting out-of-memory problems when loading larger models
> 
> @Charence Do we have access to that model? If no, could you upload the model to account named `charence`? I have one idea about what is wrong but I need a proper model to verify it. I think `Skanska federation` is the biggest model, which we have access to, but everything looks good for it.

I uploaded it into charence teamspace on your database. You should be able to see a project called blackfriairs and there's a single federation with 2 models inside :)

simplypixi commented:
@Charence fixed and rebased ;)

carmenfan commented on frontend/modules/tree/treeProcessing/transforming.ts:
```diff
@@ -13,7 +13,7 @@ const getNamespacedId = (node) => {
 const getTransformedNodeData = (node) => ({
 	_id: node._id,
-	name: node.name,
+	name: node.name || node._id,
```
@dleszcz @simplypixi Any insights? :)

dleszcz replied:
@carmenfan Ok, I'll do this

simplypixi replied:
@carmenfan
We were pretty sure it should work that way. xD

1. Why is this not handled on the BE side?
2. Should we hide list item only or get rid it of from the data?
3. What if the tree node (I mean node on the list) is a model and all of the child nodes don't have the name?

carmenfan replied:
1. Because it's not really appropriate for the BE can handle. A transformation may have multiple mesh children with no name, there's also situations where there's a transformation with a nameless mesh and other transformation children, so it's impossible for us to merge the 2 together.

2. Hide it from view, so when the user clicks on a nameless mesh, it should highlight it's parent (and any siblings)

3. only nameless meshes should be hidden. nameless transformations should be shown with `(No Name)`

dleszcz replied:
@carmenfan Done

Also rebased staging and I see that Santiago made changes in groups colors overriding. Unfortunately, he used old angularjs Tree service :/ 

I fixed this old TreeService usage and it looks like it works well with our migrated tree module, but I noticed also some watcher in _viewer.component.ts_ to update colors overrides.  We shouldn't use Angular anymore. Can you @sanmont3drepo migrate this?
By the way - why the color overriding is moved from saga to helpers? Now is impossible to use selectors in a proper way, and I had to use migration _getState()_ function.

sanmont3drepo replied:
> @carmenfan Done
> 
> Also rebased staging and I see that Santiago made changes in groups colors overriding. Unfortunately, he used old angularjs Tree service :/
> 
> I fixed this old TreeService usage and it looks like it works well with our migrated tree module, but I noticed also some watcher in _viewer.component.ts_ to update colors overrides. We shouldn't use Angular anymore. Can you @sanmont3drepo migrate this?
> By the way - why the color overriding is moved from saga to helpers? Now is impossible to use selectors in a proper way, and I had to use migration _getState()_ function.

Hi Daniel, the way I've change the groups was to make the separation between state and the UI, reacting to the changes in the state as React does. This actually allows to use selectors as the transformation of the state ( as you can see as the selectors added for groups overrides) having  one source of truth instead of changing the override list in every interaction and trying to sync searches, overrides and edition of groups which tends to have problems (that's why we've chosen redux, right?). 

**The angular part only uses the suscription to see the difference between two states and call to the appropiate unity functions in an imperative way**.

I truly tried to patch it up using calls directly in the sagas instead of the other approach but it didn't work. Both worlds kind of collide.

 This goes in hand which the thing I've told @simplypixi  long ago about having something that actually represents the viewer as a UI layer instead of making calls in the sagas to unity functions.

I can change and try to push that part of angular into the react group's component, which is not the best because it doesn't represent the actual unity viewer but I think I'd work.
 
The only potential problem that I see is the getNodesFromSharedIds if you put that in the sagas part instead of a service.

Let me think if you think that there is another way to migrate this. I have no problem in doing it.

dleszcz replied:
Ok, so let's leave it now, it works fine. We'll finish migrating it after merging these changes with #1590

dleszcz commented:
@carmenfan @sanmont3drepo 
https://github.com/3drepo/3drepo.io/pull/1535#discussion_r307687603

carmenfan commented:
@Charence will be doing another round of testing.

simplypixi commented:
> @Charence will be doing another round of testing.

@carmenfan ? :)

Charence commented:
When comparing the latest of ISSUE_1475 vs master, actions such as selection and deselection are considerably slower. Also got an error when trying to select a whole submodel.

![Aug-05-2019 17-41-18](https://user-images.githubusercontent.com/1438332/62481843-7b73fb80-b7ab-11e9-9b29-f1eabbd29dca.gif)
![Aug-05-2019 17-43-32](https://user-images.githubusercontent.com/1438332/62481854-8333a000-b7ab-11e9-8dbb-5848be7e616b.gif)

simplypixi commented:
> When comparing the latest of ISSUE_1475 vs master, actions such as selection and deselection are considerably slower. Also got an error when trying to select a whole submodel.
> 
Whaaaat? :O

Two questions:
1. Do we have access to that federation? (cc: @carmenfan )
I would like to check it.
2. @Charence which build type (production/development) did you run locally? The development build is always slower than production, so maybe that's the reason. :)

carmenfan commented:
> 1. Do we have access to that federation? (cc: @carmenfan )
>    I would like to check it.

It's the blackfriars model I uploaded into your db last time

Charence commented:
> 2\. @Charence which build type (production/development) did you run locally? The development build is always slower than production, so maybe that's the reason. :)

@simplypixi I have run it locally again using the production build type and it is still slower and has an error when I try to select a whole submodel :(

3drepo.io production:
![Aug-06-2019 11-32-02](https://user-images.githubusercontent.com/1438332/62533440-9ba0ca80-b83e-11e9-87d4-a6fde3ca0609.gif)

local production:
![Aug-06-2019 11-33-08](https://user-images.githubusercontent.com/1438332/62533455-a0fe1500-b83e-11e9-80b5-e4f9c45291c2.gif)

simplypixi commented:
> > 2. @Charence which build type (production/development) did you run locally? The development build is always slower than production, so maybe that's the reason. :)
> 
> @simplypixi I have run it locally again using the production build type and it is still slower and has an error when I try to select a whole submodel :(
> 
> 3drepo.io production:
> ![Aug-06-2019 11-32-02](https://user-images.githubusercontent.com/1438332/62533440-9ba0ca80-b83e-11e9-87d4-a6fde3ca0609.gif)
> 
> local production:
> ![Aug-06-2019 11-33-08](https://user-images.githubusercontent.com/1438332/62533455-a0fe1500-b83e-11e9-80b5-e4f9c45291c2.gif)

@carmenfan @Charence could you upload this model to our database?

carmenfan commented:
@simplypixi  that should be the blackfriars fed I uploaded for you

simplypixi commented:
> @simplypixi that should be the blackfriars fed I uploaded for you

Ohh, ok :)

carmenfan commented:
@simplypixi updated unity

simplypixi commented:
> Also got an error when trying to select a whole submodel.

@Charence fixed

simplypixi commented:
@carmenfan I migrated Santiago's change (watcher - which we discussed in Model/Viewer PR) and resolved conflicts (I had conflicts related to unity files so maybe next build will be necessary).
I'm still debugging low performance for BlackFrairs model.

simplypixi commented:
@carmenfan I returned to this branch after preparing fix for security issues and I'm getting unity error (see screen below). Do you have any idea what is going? :O
![Screenshot at Aug 20 14-23-55](https://user-images.githubusercontent.com/5683490/63346896-41335e00-c356-11e9-9bb9-3ad760af2743.png)

carmenfan commented:
@simplypixi I just committed another unity update on this branch, see if it works?

(This has the latest staging changes with the unity loader changes, so if you merge staging in, take this as latest)

simplypixi commented:
> @simplypixi I just committed another unity update on this branch, see if it works?
> 
> (This has the latest staging changes with the unity loader changes, so if you merge staging in, take this as latest)

It works. Thanks! ;)

simplypixi commented:
@Charence I found one bottleneck and fixed it. Could you verify if currently selection works faster on your side?

Charence commented:
> > Also got an error when trying to select a whole submodel.
> 
> @Charence fixed

I still seem to be getting problems with this. I now get a blank screen instead of an error dialog. I tried restarting chrome. Does it always work smoothly for you?

![Aug-27-2019 15-01-36](https://user-images.githubusercontent.com/1438332/63778356-8b7b8880-c8dc-11e9-99d4-bbb65598cca1.gif)

carmenfan commented:
@simplypixi Is there any reason why we made the model loading sequential? (Doesn't seems like a change on this branch particularly, but I just noticed it. It seems like we don't even fetch the viewer until the tree is fetched from the server. For large models the trees can be in 10s or 100s MBs, causing the user to not get any response for minutes before they get the prompt on loading viewer/loading model etc.

simplypixi commented:
> For large mod



> @simplypixi Is there any reason why we made the model loading sequential? (Doesn't seems like a change on this branch particularly, but I just noticed it. It seems like we don't even fetch the viewer until the tree is fetched from the server. For large models the trees can be in 10s or 100s MBs, causing the user to not get any response for minutes before they get the prompt on loading viewer/loading model etc.

Hmm, it sounds weird. Could paste here some code sample or gif? Is it happening on staging?

carmenfan commented:
I tried to make a gif but it's massive due to the amount of frames... lol

But if you look at the network tab this is what it's doing it seems:

- Load full tree + tree path
- once finished, load sub model full tree (this shouldn't need to wait for tree path!)
- once finished, load unity game + models. (This shouldn't need to wait for tree to finish loading to happen!)

So seems like there's some unnecessary async/await situation here.


I actually just realised that I've only tried this on staging, not this branch. So it could be that this is resolved on this branch. None the less, it's worth double checking!

simplypixi commented:
> I actually just realised that I've only tried this on staging, not this branch. So it could be that this is resolved on this branch. None the less, it's worth double checking!

It works on this branch because I rewrote loading logic a bit.
Should we fix it also on staging branch or can we wait to merge this PR? :)

carmenfan commented:
@simplypixi How much work would it be to do the fix in staging? It's likely that we'll be doing the next release without this PR, so it would be good to have it in if it's not too much effort.

simplypixi commented:
> @simplypixi I still seem to be getting problems with this. I now get a blank screen instead of an error dialog. I tried restarting chrome. Does it always work smoothly for you?

@Charence 
The previous bug was fixed. It's another one. I think it's related to this model.
The last time I resigned debugging the bug on "Blackfriars" model and I did it on smaller one - because that model is huge and reloading the app after each change was taking too much time.


> I tried to make a gif but it's massive due to the amount of frames... lol
> 
> But if you look at the network tab this is what it's doing it seems:
> 
> * Load full tree + tree path
> * once finished, load sub model full tree (this shouldn't need to wait for tree path!)
> * once finished, load unity game + models. (This shouldn't need to wait for tree to finish loading to happen!)
> 
> So seems like there's some unnecessary async/await situation here.
> 
> I actually just realised that I've only tried this on staging, not this branch. So it could be that this is resolved on this branch. None the less, it's worth double checking!

So, should we fix it also on staging or can we wait for merging this PR?

> @simplypixi How much work would it be to do the fix in staging? It's likely that we'll be doing the next release without this PR, so it would be good to have it in if it's not too much effort.

@carmenfan I'm in the middle of investigating this bug. I'll let you know in about 15mins ;)

simplypixi commented:
> > @simplypixi How much work would it be to do the fix in staging? It's likely that we'll be doing the next release without this PR, so it would be good to have it in if it's not too much effort.
> 
> @carmenfan I'm in the middle of investigating this bug. I'll let you know in about 15mins ;)

@carmenfan I won't be a quick fix. :/ 
But I'm working on it. ;)

simplypixi commented:
> > > @simplypixi How much work would it be to do the fix in staging? It's likely that we'll be doing the next release without this PR, so it would be good to have it in if it's not too much effort.
> > 
> > 
> > @carmenfan I'm in the middle of investigating this bug. I'll let you know in about 15mins ;)
> 
> @carmenfan I won't be a quick fix. :/
> But I'm working on it. ;)

@carmenfan I have a fix! :O
(A solution was very simple - see commit from my new PR) ;)

dleszcz commented:
> > > Also got an error when trying to select a whole submodel.
> > 
> > 
> > @Charence fixed
> 
> @simplypixi I still seem to be getting problems with this. I now get a blank screen instead of an error dialog. I tried restarting chrome. Does it always work smoothly for you?
> 
> ![Aug-27-2019 15-01-36](https://user-images.githubusercontent.com/1438332/63778356-8b7b8880-c8dc-11e9-99d4-bbb65598cca1.gif)
> 
> Also, in comparison to the version on 3drepo.io, selection and deselection is still noticeably slower.

@Charence Done

Charence commented:
> @Charence Done

@dleszcz I can see the problem with submodel selection is fixed, but please see the comparison between the latest of ISSUE_1475 (left) vs master (right). I hope improvements in selection and deselection can still be made.

![Aug-30-2019 16-26-53](https://user-images.githubusercontent.com/1438332/64032829-30979a80-cb43-11e9-9ddb-ded016e344f5.gif)

carmenfan commented:
> > @Charence Done
> 
> @dleszcz I can see the problem with submodel selection is fixed, but please see the comparison between the latest of ISSUE_1475 (left) vs master (right). I hope improvements in selection and deselection can still be made.
> 
> ![Aug-30-2019 16-26-53](https://user-images.githubusercontent.com/1438332/64032829-30979a80-cb43-11e9-9ddb-ded016e344f5.gif)

@simplypixi @dleszcz 
As you guys are aware, speed is a major factor of this component. If it's not near the speed of the current implementation, we can't release it.

It would save a lot of going back and forth if you guys can benchmark the experience against dev/production with the same model and see if it fast enough before telling us to check it. As that is exactly what we'd be doing to tell you if it's good enough 😆

dleszcz commented:
> > > @Charence Done
> > 
> > 
> > @dleszcz I can see the problem with submodel selection is fixed, but please see the comparison between the latest of ISSUE_1475 (left) vs master (right). I hope improvements in selection and deselection can still be made.
> > ![Aug-30-2019 16-26-53](https://user-images.githubusercontent.com/1438332/64032829-30979a80-cb43-11e9-9ddb-ded016e344f5.gif)
> 
> @simplypixi @dleszcz
> As you guys are aware, speed is a major factor of this component. If it's not near the speed of the current implementation, we can't release it.
> 
> It would save a lot of going back and forth if you guys can benchmark the experience against dev/production with the same model and see if it fast enough before telling us to check it. As that is exactly what we'd be doing to tell you if it's good enough 😆

Ok, then I will work more on performance.

dleszcz commented:
Hello @carmenfan, one question :)
Have you tested it on the production or development build?

carmenfan commented:
@dleszcz I've always tested it on dev build as prod build takes way too long.

Are you suggesting the prod build matches the speed of staging? :)

dleszcz commented:
> @dleszcz I've always tested it on dev build as prod build takes way too long.
> 
> Are you suggesting the prod build matches the speed of staging? :)

That is, theoretically the production build should be faster than the development one.
Also I have added improvements, and in my opinion the times on this huge model are similar on staging and production build on my local machine. But I will still try to reduce this selection time ;)

dleszcz commented:
Hello @carmen @Charence 

I spent some time optimizing nodes selection. I managed to reduce the time a little bit more. Now the action _selectNodes_ lasts on average ~2.3s (dev build) and ~1.8s (prod build). Before it lasted about 5-6 seconds. Tested on _Blackfriars_.

Please check and let me know if that's enough for you :)

carmenfan commented on frontend/modules/tree/treeProcessing/processing.ts:
```diff
@@ -0,0 +1,419 @@
+	}
+
+	public selectNodes = ({ nodesIds = [], ...extraData }) => {
+		const visibleNodesIds = nodesIds.filter((nodeId) => this.visibilityMap[nodeId] !== VISIBILITY_STATES.INVISIBLE);
```
this line and L87  should merge together into a single loop.

there's 3  filters and getNodesByIds is another loop, resulting in O(3n) when we could do it in O(n).

Also the filter at L87 is redundant as L81 already done it?

carmenfan commented on frontend/modules/tree/treeProcessing/processing.ts:
```diff
@@ -0,0 +1,419 @@
+			this.visibilityMap[node._id] !== VISIBILITY_STATES.INVISIBLE
+		);
+
+		if (!extraData.skipChildren) {
```
What is the purpose of getting all the children of the nodes to select?

dleszcz replied:
Honestly I don't remember because it was implemented by Mariusz ;) How should it be in your opinion? 

https://github.com/3drepo/3drepo.io/pull/1535#discussion_r322247370

carmenfan commented on frontend/modules/tree/treeProcessing/processing.ts:
```diff
@@ -0,0 +1,419 @@
+		}
+
+		this.handleToSelect(nodes);
+		const highlightedObjects = this.getMeshesByNodes(nodes);
```
since nodes is now compactNodes, we're getting meshes by nodes with the original nodes and also its children. This isn't necessary as the original nodes would've given you the same list of highlightedObjects, so looping through its children is redundant.

dleszcz replied:
@carmenfan Are you sure? Because I tried to do this without children and then I don't get any highlighted objects

carmenfan commented on frontend/modules/tree/treeProcessing/processing.ts:
```diff
@@ -0,0 +1,419 @@
+		return this.nodesList.slice(nodeIndex + 1, nodeIndex + node.deepChildrenNumber + 1);
+	}, (node) => node._id);
+
+	public getParents = memoize((node = {}) => {
```
instead of doing the while loop, you should just use `idToPath` from `tree_path.json`

for e.g. node `A` would have an entry like `B__C__A`

B and C are A's parents.

carmenfan commented on frontend/modules/tree/treeProcessing/processing.ts:
```diff
@@ -0,0 +1,419 @@
+
+				const meshes = node.type === NODE_TYPES.MESH
+					? [node._id]
+					: this.meshesByModelId[node._id];
```
`meshesByModelId` is badly named. We're accessing it via node ID here.

carmenfan commented on frontend/modules/tree/treeProcessing/processing.ts:
```diff
@@ -0,0 +1,419 @@
+					: this.meshesByModelId[node._id];
+
+				if (meshes) {
+					const meshesByNodesIndex = meshesByNodesIndexes[node.namespacedId];
```
plural of Index is Indices 😆

dleszcz commented on frontend/modules/tree/treeProcessing/treeProcessing.ts:
```diff
@@ -0,0 +1,46 @@
+
+	public isolateNodes = (payload) => this.processing.isolateNodes(payload);
+
+	public clearSelected = () => this.processing.clearCurrentlySelected();
```
I will also look at this, thanks :)

carmenfan commented:
@dleszcz 
![image](https://user-images.githubusercontent.com/11945337/64975187-4ed0fa80-d8a6-11e9-9f01-e12b5d63071d.png)

It's completely broken right now..

dleszcz commented:
> @dleszcz
> ![image](https://user-images.githubusercontent.com/11945337/64975187-4ed0fa80-d8a6-11e9-9f01-e12b5d63071d.png)
> 
> It's completely broken right now..
> 
> Edit:
> If you fix this and merge staging into this branch (as there's a lot of conflicts). I'll take another look.

Hello @carmenfan 
I checked it out. It broke after the last Stantiago's commit (0fef964). (btw: the tree doesn't load either)

I switched on my last commit and selecting works ok. Should I check it or @sanmont3drepo ? ;)

sanmont3drepo commented:
> > @dleszcz
> > ![image](https://user-images.githubusercontent.com/11945337/64975187-4ed0fa80-d8a6-11e9-9f01-e12b5d63071d.png)
> > It's completely broken right now..
> > Edit:
> > If you fix this and merge staging into this branch (as there's a lot of conflicts). I'll take another look.
> 
> Hello @carmenfan
> I checked it out. It broke after the last Stantiago's commit ([0fef964](https://github.com/3drepo/3drepo.io/commit/0fef9647cf64d9df3126a18ba86972afc658fcd2)). (btw: the tree doesn't load either)
> 
> I switched on my last commit and everything works ok. Should I check it? or @sanmont3drepo ? ;)

Fixed in 0e66660ab771854291f17e16c3a3b1354b6fe7bc

carmenfan commented:
@simplypixi 
![image](https://user-images.githubusercontent.com/11945337/65898284-130b5a00-e3a9-11e9-8f9c-82cf887d2e23.png)

Should loading.worker.ts be removed? I don't think we're using it anymore and it's basically a copy (but out of date) version of what's in `transforming.ts`?

simplypixi commented:
> @simplypixi
> ![image](https://user-images.githubusercontent.com/11945337/65898284-130b5a00-e3a9-11e9-8f9c-82cf887d2e23.png)
> 
> Should loading.worker.ts be removed? I don't think we're using it anymore and it's basically a copy (but out of date) version of what's in `transforming.ts`?

Yes! It could be removed. ;)

carmenfan commented:
@simplypixi  @dleszcz 
Seems like we lost the functionality of showing the model loading progress on the new message at the bottom. It's not disabling the unity's red loading bar up top and the model loading message is stuck at 0%. 

Can you guys take a look?

carmenfan commented:
> @simplypixi @dleszcz
> Seems like we lost the functionality of showing the model loading progress on the new message at the bottom. It's not disabling the unity's red loading bar up top and the model loading message is stuck at 0%.
> 
> Can you guys take a look?

Actually it seems to be working now with a new unity build... Maybe it didn't have the change!

carmenfan commented:
@simplypixi I'm currently seeing this problem when we have this branch on a dev server

- We deployed this branch at commit A
- We then did further fixes to the branch (let's call this commit B)
- We then deployed commit B onto the server

When I load the page i'm seeing commit A's version. Then I did a hard refresh and it becomes commit B.

This sounds very much like the service worker loading the cache version, which is why we had that "there's a new version of 3D Repo" dialog that ask if the user wants to refresh to get the latest version. 

If this is what's happening, can we make it so it somehow doesn't load the cache version if it's out of date? if not, can we revive that dialog so the user knows and is given the option to refresh?

carmenfan commented:
> > @simplypixi @dleszcz
> > Seems like we lost the functionality of showing the model loading progress on the new message at the bottom. It's not disabling the unity's red loading bar up top and the model loading message is stuck at 0%.
> > Can you guys take a look?
> 
> Actually it seems to be working now with a new unity build... Maybe it didn't have the change!

@simplypixi  I think it's the cache that's confusing me, but it's definitely not showing the % for both loading viewer and loading model.. can you please take a look 😆

carmenfan commented:
@simplypixi @dleszcz I'm merging the tree in!

There's some problems with service worker behaviour and model loading % prompt, which I will create separate issues for
